src/nvenc/nvenc_base.cpp

```diff
@@ -0,0 +1,323 @@
+      return "P2";
+    }
+    if (guid == NV_ENC_PRESET_P3_GUID) {
+      return "P3";
+    }
+    if (guid == NV_ENC_PRESET_P4_GUID) {
+      return "P4";
+    }
+    if (guid == NV_ENC_PRESET_P5_GUID) {
+      return "P5";
+    }
+    if (guid == NV_ENC_PRESET_P6_GUID) {
+      return "P6";
+    }
+    if (guid == NV_ENC_PRESET_P7_GUID) {
+      return "P7";
+    }
+    return "Unknown";
+  }
+
+}  // namespace
+
+namespace nvenc {
+
+  nvenc_base::nvenc_base(NV_ENC_DEVICE_TYPE device_type, void *device, uint32_t max_width, uint32_t max_height, NV_ENC_BUFFER_FORMAT buffer_format):
+      device_type(device_type),
+      device(device),
+      max_width(max_width),
+      max_height(max_height),
+      buffer_format(buffer_format) {
+  }
+
+  nvenc_base::~nvenc_base() {
+    // Use destroy_base_resources() instead
+  }
+
+  bool
+  nvenc_base::create_encoder(const nvenc_config &config, const video::config_t &client_config, const nvenc_colorspace &colorspace) {
+    if (encoder) return false;
+    if (!nvenc && !init_library()) return false;
+
+    NV_ENC_OPEN_ENCODE_SESSION_EX_PARAMS session_params = { NV_ENC_OPEN_ENCODE_SESSION_EX_PARAMS_VER };
+    session_params.device = device;
+    session_params.deviceType = device_type;
+    session_params.apiVersion = NVENCAPI_VERSION;
+    if (nvenc->nvEncOpenEncodeSessionEx(&session_params, &encoder) != NV_ENC_SUCCESS) {
+      BOOST_LOG(error) << "NvEncOpenEncodeSessionEx failed";
+      return false;
+    }
+
+    width = client_config.width;
+    height = client_config.height;
+
+    NV_ENC_INITIALIZE_PARAMS init_params = { NV_ENC_INITIALIZE_PARAMS_VER };
+    init_params.maxEncodeWidth = max_width;
+    init_params.maxEncodeHeight = max_height;
+    init_params.presetGUID = quality_preset_guid_from_number(config.quality_preset);
+    init_params.tuningInfo = NV_ENC_TUNING_INFO_ULTRA_LOW_LATENCY;
+    init_params.enablePTD = 1;
+
+    init_params.encodeWidth = width;
+    init_params.darWidth = width;
+    init_params.encodeHeight = height;
+    init_params.darHeight = height;
+    init_params.frameRateNum = client_config.framerate;
+    init_params.frameRateDen = 1;
+
+    switch (client_config.videoFormat) {
+      case 0:
+        // H.264
+        init_params.encodeGUID = NV_ENC_CODEC_H264_GUID;
+        break;
+
+      case 1:
+        // HEVC
+        init_params.encodeGUID = NV_ENC_CODEC_HEVC_GUID;
+        break;
+    }
+
+    NV_ENC_PRESET_CONFIG preset_config = { NV_ENC_PRESET_CONFIG_VER, { NV_ENC_CONFIG_VER } };
+    if (nvenc->nvEncGetEncodePresetConfigEx(encoder, init_params.encodeGUID, init_params.presetGUID, init_params.tuningInfo, &preset_config) != NV_ENC_SUCCESS) {
+      BOOST_LOG(error) << "NvEncGetEncodePresetConfigEx failed: " << nvenc->nvEncGetLastErrorString(encoder);
+      return false;
+    }
+
+    NV_ENC_CONFIG enc_config = preset_config.presetCfg;
+    enc_config.profileGUID = NV_ENC_CODEC_PROFILE_AUTOSELECT_GUID;  // TODO: select specific profile?

```

Comment:
Yeah, we should select specific profiles. The specific values should be:
- `videoFormat == 0` -> `NV_ENC_H264_PROFILE_HIGH_GUID `
- `videoFormat == 1 && dynamicRange == 0` -> `NV_ENC_HEVC_PROFILE_MAIN_GUID`
- `videoFormat == 1 && dynamicRange == 1` -> `NV_ENC_HEVC_PROFILE_MAIN10_GUID`

Reply:
H.264 Main or Baseline should work, but I think we should stick to H.264 High since that's what all the other encoders do. We manually specify the profile for the FFmpeg-based encoders to ensure we always get what we expect.

Reply:
Forced H264 to High, the rest should be fine on auto.